### ISSUE TITLE
_NoFill results in junk loading

### ISSUE THREAD

**Issue by znicholls** [SciTools/iris#3385]
**tl;dr**

If a variable has `_NoFill = "true"` then loaded data doesn't mask properly. There was some discussion of this in #2748 (where I think it was decided that it wasn't a high priority..?). 

**Correction**

Originally I thought some CMIP6 data has `_NoFill = "true"`. I'm not actually sure how this happened anymore.

**Demonstration**

Below is a complete minimal working example (with scripts to generate the test data at the end, thanks to http://www.ceda.ac.uk/static/media/uploads/ncas-reading-2015/11_create_netcdf_python.pdf for guide on how to generate these files).

```python
>>> import iris
>>> iris.__version__
'2.2.1dev0'
```

(This could be put into a unit test)

```sh
$ ncdump -sh no_fill_value.nc 
netcdf no_fill_value {
dimensions:
	lat = 73 ;
	lon = 144 ;
	time = UNLIMITED ; // (5 currently)
variables:
	float lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "Latitude" ;
		lat:units = "degrees_north" ;
		lat:axis = "Y" ;
		lat:_Storage = "contiguous" ;
		lat:_Endianness = "little" ;
		lat:_NoFill = "true" ;
	float lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "Longitude" ;
		lon:units = "degrees_east" ;
		lon:axis = "X" ;
		lon:_Storage = "contiguous" ;
		lon:_Endianness = "little" ;
		lon:_NoFill = "true" ;
	double time(time) ;
		time:units = "hours since 0001-01-01 00:00:00" ;
		time:calendar = "gregorian" ;
		time:_Storage = "chunked" ;
		time:_ChunkSizes = 512 ;
		time:_Endianness = "little" ;
		time:_NoFill = "true" ;
	float temp(time, lat, lon) ;
		temp:units = "K" ;
		temp:_Storage = "chunked" ;
		temp:_ChunkSizes = 1, 73, 144 ;
		temp:_Endianness = "little" ;
		temp:_NoFill = "true" ;

// global attributes:
		:_NCProperties = "version=1|netcdflibversion=4.6.1|hdf5libversion=1.10.2" ;
		:_SuperblockVersion = 0 ;
		:_IsNetcdf4 = 1 ;
		:_Format = "netCDF-4" ;
}
```

```python
>>> no_fill_value = iris.load_cube("no_fill_value.nc")
>>> no_fill_value.data.mask.any()
False  # None of the data is masked
>>> no_fill_value.data
masked_array(
  data=[[[0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.54024243, 0.6060777 , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.10345457, 0.9351688 , ..., 0.        ,
          0.        , 0.        ],
         ...,
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ]],

        [[0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.05158424, 0.33022666, ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.7684858 , 0.93951946, ..., 0.        ,
          0.        , 0.        ],
         ...,
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ]],

        [[0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.22425927, 0.8433138 , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.33034268, 0.40054017, ..., 0.        ,
          0.        , 0.        ],
         ...,
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ]],

        [[0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.58714134, 0.67210656, ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.31482515, 0.35548896, ..., 0.        ,
          0.        , 0.        ],
         ...,
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ]],

        [[0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.38941053, 0.4985403 , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.32742476, 0.2148281 , ..., 0.        ,
          0.        , 0.        ],
         ...,
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ],
         [0.        , 0.        , 0.        , ..., 0.        ,
          0.        , 0.        ]]],
  mask=False,  # this is wrong
  fill_value=1e+20,
  dtype=float32)
```

If `_NoFill` is not included, everything masks properly.

```sh
$ ncdump -sh with_fill_value.nc 
netcdf with_fill_value {
dimensions:
	lat = 73 ;
	lon = 144 ;
	time = UNLIMITED ; // (5 currently)
variables:
	float lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "Latitude" ;
		lat:units = "degrees_north" ;
		lat:axis = "Y" ;
		lat:_Storage = "contiguous" ;
		lat:_Endianness = "little" ;
	float lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "Longitude" ;
		lon:units = "degrees_east" ;
		lon:axis = "X" ;
		lon:_Storage = "contiguous" ;
		lon:_Endianness = "little" ;
	double time(time) ;
		time:units = "hours since 0001-01-01 00:00:00" ;
		time:calendar = "gregorian" ;
		time:_Storage = "chunked" ;
		time:_ChunkSizes = 512 ;
		time:_Endianness = "little" ;
	float temp(time, lat, lon) ;
		temp:units = "K" ;
		temp:_Storage = "chunked" ;
		temp:_ChunkSizes = 1, 73, 144 ;
		temp:_Endianness = "little" ;

// global attributes:
		:_NCProperties = "version=1|netcdflibversion=4.6.1|hdf5libversion=1.10.2" ;
		:_SuperblockVersion = 0 ;
		:_IsNetcdf4 = 1 ;
		:_Format = "netCDF-4" ;
}
```

```python
>>> with_fill_value = iris.load_cube("with_fill_value.nc")
>>> with_fill_value.data.mask.any()
True  # The data is masked
>>> with_fill_value.data
masked_array(
  data=[[[--, --, --, ..., --, --, --],
         [--, 0.8181109428405762, 0.959919273853302, ..., --, --, --],
         [--, 0.14339472353458405, 0.9362871050834656, ..., --, --, --],
         ...,
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --]],

        [[--, --, --, ..., --, --, --],
         [--, 0.5713177919387817, 0.7257010340690613, ..., --, --, --],
         [--, 0.41555869579315186, 0.39882218837738037, ..., --, --, --],
         ...,
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --]],

        [[--, --, --, ..., --, --, --],
         [--, 0.3753425180912018, 0.47521865367889404, ..., --, --, --],
         [--, 0.9013395309448242, 0.7477187514305115, ..., --, --, --],
         ...,
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --]],

        [[--, --, --, ..., --, --, --],
         [--, 0.6569557785987854, 0.9291857481002808, ..., --, --, --],
         [--, 0.7672317028045654, 0.5529156923294067, ..., --, --, --],
         ...,
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --]],

        [[--, --, --, ..., --, --, --],
         [--, 0.985698401927948, 0.8145892024040222, ..., --, --, --],
         [--, 0.14124305546283722, 0.7355625629425049, ..., --, --, --],
         ...,
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --],
         [--, --, --, ..., --, --, --]]],
  mask=[[[ True,  True,  True, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         ...,
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True]],

        [[ True,  True,  True, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         ...,
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True]],

        [[ True,  True,  True, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         ...,
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True]],

        [[ True,  True,  True, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         ...,
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True]],

        [[ True,  True,  True, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         [ True, False, False, ...,  True,  True,  True],
         ...,
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True],
         [ True,  True,  True, ...,  True,  True,  True]]],
  fill_value=9.96921e+36,
  dtype=float32)
```

**Generating test data**

```python
# generate_test_data.py
import numpy as np
from netCDF4 import Dataset
from numpy.random import uniform

def create_example_dataset(out_path, no_fill):
    dataset = Dataset(out_path, "w")
    if no_fill:
        dataset.set_fill_off()

    lat = dataset.createDimension('lat', 73)
    lon = dataset.createDimension('lon', 144)
    time = dataset.createDimension('time', None)
    
    latitudes = dataset.createVariable('lat', np.float32, ('lat',))
    longitudes = dataset.createVariable('lon', np.float32, ('lon',))
    times = dataset.createVariable('time', np.float64, ('time',))
    temp = dataset.createVariable('temp', np.float32, ("time", "lat", "lon"))

    latitudes.standard_name = 'latitude'
    latitudes.long_name = 'Latitude'
    latitudes.units = 'degrees_north'
    latitudes.axis = 'Y'

    longitudes.standard_name = 'longitude'
    longitudes.long_name = 'Longitude'
    longitudes.units = 'degrees_east'
    longitudes.axis = 'X'

    temp.units = 'K'

    times.units = 'hours since 0001-01-01 00:00:00'
    times.calendar = 'gregorian'

    lats = np.arange(-90,91,2.5)
    lons = np.arange(-180,180,2.5)
    latitudes[:] = lats
    longitudes[:] = lons

    nlats = len(dataset.dimensions['lat'])
    nlons = len(dataset.dimensions['lon'])
    temp[0:5,1:3,1:4] = uniform(size=(5,2,3))

    times[:] = [0, 1, 2, 3, 4]

    dataset.close()

create_example_dataset("no_fill_value.nc", True)
create_example_dataset("with_fill_value.nc", False)
```

```sh
python generate_test_data.py
```

**Comment by pp-mo** [comment]
Hi @znicholls
Thanks for providing such a clear example.  Awesome! :bouquet: :100: 

I looked into the handling of this in Iris, netCDF4-python and the underlying netcdf libraries.
Unfortunately, my understanding is that this behaviour is ***not*** wrong, according to the current conventions.

I used your code to generate the files + investigated ...

The 'with_fill_value.nc" file is written with 'normal' behaviour, equivalent to `dataset.set_fill_on()`.  The unwritten data points are thus initialised to the netcdf default fill-value for the type, which is just what netcdf defines.  If you look the underlying "raw" data values read from the file (the masked-array.data), you get ...
```
>>> print(ds.var['temp'][...].mask.any())
True
>>> print(ds.var['temp'][0, :4, :4])
[[-- -- -- --]
 [-- 0.11086352914571762 0.4374963641166687 0.8171334862709045]
 [-- 0.5340825915336609 0.03521599993109703 0.5274709463119507]
 [-- -- -- --]]
>>> print(ds.var['temp'][0, :4, :4].data)
[[9.9692100e+36 9.9692100e+36 9.9692100e+36 9.9692100e+36]
 [9.9692100e+36 1.1086353e-01 4.3749636e-01 8.1713349e-01]
 [9.9692100e+36 5.3408259e-01 3.5216000e-02 5.2747095e-01]
 [9.9692100e+36 9.9692100e+36 9.9692100e+36 9.9692100e+36]]
```

Then, the "no_fill_value.nc" file is written with `dataset.set_fill_off()`, which means there is a `_NoFill= "true"` attribute attached to each variable, which can be seen in the dump.
That means the unwritten points are uninitialised.  In this case they seem to be zeroes...
```
>>> print(ds.var['temp'][...].mask.any())
False
>>> print(ds.var['temp'][0, :4, :4])
[[0.         0.         0.         0.        ]
 [0.         0.92597336 0.19297834 0.7853463 ]
 [0.         0.15583207 0.7349046  0.69109625]
 [0.         0.         0.         0.        ]]
>>> print(ds.var['temp'][0, :4, :4].data)
[[0.         0.         0.         0.        ]
 [0.         0.92597336 0.19297834 0.7853463 ]
 [0.         0.15583207 0.7349046  0.69109625]
 [0.         0.         0.         0.        ]]
```
But, because none of the software actually treats a '_NoFill' parameter, this is "just" a coincidence: The unwritten points "happen" to be 0, which does not match the (default) fill-value.

If, when creating the "temp" variable, you ***also*** set `fill_value=0`, then the variable gets a `_FillValue = 0.` attribute --  
and then these same unwritten points will ***also*** read back as masked...
```
>>> print([(att, getattr(var, att)) for att in ds.var['temp'].ncattrs()])
[(u'_FillValue', 0.0), (u'units', u'K')]
>>> print(ds.var['temp'][...].mask.any())
True
>>> print(ds.var['temp'][0, :4, :4])
[[-- -- -- --]
 [-- 0.9227437973022461 0.3860781788825989 0.5692782998085022]
 [-- 0.5689221024513245 0.5872024297714233 0.833384096622467]
 [-- -- -- --]]
>>> print(ds.var['temp'][0, :4, :4].data)
[[0.         0.         0.         0.        ]
 [0.         0.9227438  0.38607818 0.5692783 ]
 [0.         0.5689221  0.5872024  0.8333841 ]
 [0.         0.         0.         0.        ]]

```
I.E. there is no way of avoiding that 'some' possible value is interpreted as a missing value, there is always a comparison check.

Do you think that a `_NoFill="true"` attribute **should indicate that the raw data are to be treated as unmasked : all values valid, no fill-value check to be made ?**
I have heard something similar elsewhere, but I don't think this is a correct intention for the "_NoFill" attribute.

The [netcdf 4.7.1. docs](https://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html) say that _NoFill only reflects what was done when writing.
The netcdf 1.5.1 docs on the [set_fill_on/off calls](https://unidata.github.io/netcdf4-python/netCDF4/index.html#netCDF4.Dataset.set_fill_off) seems consistent with that.  Likewise the per-variable 'fill_value=False" option in [createVariable](https://unidata.github.io/netcdf4-python/netCDF4/index.html#netCDF4.Dataset.createVariable) does the same thing on an individual variable
"If fill_value is set to False, then the variable is not pre-filled.".

If there *were* an accepted convention to assert that all points are unmasked and no possible value means "missing data", I think that would be a great thing and we would definitely adopt it.  
But I don't think it exists (yet).

**Comment by znicholls** [comment]
Thanks for the detailed response @pp-mo ! In the meantime I worked out that the use of `_NoFill = "true"` is actually the result of a `cdo -selyear` operation with cdo 1.9.7.1 and doesn't appear in the raw CMIP6 output, apologies. Hence I'm not sure if this is such a priority any more (I can raise a bug in cdo if you think it's worth it...). 

Nonetheless, if you have time it'd be great to get your thoughts on the below.

I think I may be missing something as the current behaviour seems odd to me. What I was hoping was that the file which had `_NoFill = "true"` would come back with masked data nonetheless. The unit test would look something like this:

**Test file**

```sh
$ ncdump -sh no_fill_value.nc 
netcdf no_fill_value {
dimensions:
	lat = 73 ;
	lon = 144 ;
	time = UNLIMITED ; // (5 currently)
variables:
	float lat(lat) ;
		lat:standard_name = "latitude" ;
		lat:long_name = "Latitude" ;
		lat:units = "degrees_north" ;
		lat:axis = "Y" ;
		lat:_Storage = "contiguous" ;
		lat:_Endianness = "little" ;
		lat:_NoFill = "true" ;
	float lon(lon) ;
		lon:standard_name = "longitude" ;
		lon:long_name = "Longitude" ;
		lon:units = "degrees_east" ;
		lon:axis = "X" ;
		lon:_Storage = "contiguous" ;
		lon:_Endianness = "little" ;
		lon:_NoFill = "true" ;
	double time(time) ;
		time:units = "hours since 0001-01-01 00:00:00" ;
		time:calendar = "gregorian" ;
		time:_Storage = "chunked" ;
		time:_ChunkSizes = 512 ;
		time:_Endianness = "little" ;
		time:_NoFill = "true" ;
	float temp(time, lat, lon) ;
		temp:units = "K" ;
		temp:_Storage = "chunked" ;
		temp:_ChunkSizes = 1, 73, 144 ;
		temp:_Endianness = "little" ;
		temp:_NoFill = "true" ;

// global attributes:
		:_NCProperties = "version=1|netcdflibversion=4.6.1|hdf5libversion=1.10.2" ;
		:_SuperblockVersion = 0 ;
		:_IsNetcdf4 = 1 ;
		:_Format = "netCDF-4" ;
}
```

**Test function**

```python
def test_load_no_fill():
    no_fill_value = iris.load_cube("no_fill_value.nc")
    assert no_fill_value.data.mask.any()
```

i.e. we would be testing that iris reads in a mask, even if we have `_NoFill = "true"`.

At the moment we have 

```python
>>> no_fill_value = iris.load_cube("no_fill_value.nc")
>>> no_fill_value.data.mask.any()
False
```

i.e. all values are treated as valid if `_NoFill = "true"`.

**Comment by znicholls** [comment]
@pp-mo one more question, is there an easy way to check if `_NoFill = "true"` when reading data? I'm going around in circles trying to work out how to do it with netCDF4

**Comment by znicholls** [comment]
Doing some more digging, it looks like even `ncdump` can't tell which points to mask if `_NoFill = "true"`. So it looks like there's no way to handle this any better than it currently is? My only thought would be to throw a warning if `_NoFill = "true" if it's even possible to check for `_NoFill = "true"`.

**Comment by pp-mo** [comment]
Hi @znicholls do you still have an interest here ??
If not we should close, I think.

> @pp-mo one more question, is there an easy way to check if _NoFill = "true" when reading data? I'm going around in circles trying to work out how to do it with netCDF4

Regarding the "_NoFill", this setting won't be visible through an Iris load.
I can't work out how to see it via netcdf4-python, either.  It seems to be entirely hidden away.
To see it as part of an ncdump you must use '-s', as you are.  I note that the helpstring for that says "Output special (virtual) attributes", so perhaps something special is happening here - it certainly does not appear in the variable.ncattrs(), or to var.getncattr().

As far as the netcdf4-python / C library is concerned, if I understand correctly :  The 'no fill' is a dataset-level setting : however the file itself does not have a 'no fill' setting, it is just an operating mode of the library.  When this is enabled, it adds a '_NoFill = true' attribute when creating a variable.  I think that is the only actual difference stored in the file.

**Comment by znicholls** [comment]
> If not we should close, I think. ?

I think so. As much as it's frustrating that there seems to be no way to even warn users that their might be full of random numbers, I can't see a way forward (and have no idea who else to ask) so am happy to close. If someone stumbles upon this with a solution we can re-open then.

Cheers for looking in @pp-mo !

**Comment by pp-mo** [comment]
> 

P.S. it explains it here : https://www.unidata.ucar.edu/software/netcdf/workshops/2011/utilities/SpecialAttributes.html

The "_NoFill" attribute does not really exist, but there is a real property of the variable that is set at its creation.  It can be read with the [nc_var_inq_fill() function](https://www.unidata.ucar.edu/software/netcdf/docs/group__variables.html#gaf4a2635933f4b58a5cc2025e4e2e120f).
There doesn't seem to be any support for this via the python interface.

**Comment by pp-mo** [comment]
Thanks for your understanding @znicholls  !

**Comment by pp-mo** [comment]
Afterthought to this : #3438
